src/Compilers/CSharp/Portable/Compiler/TypeCompilationState.cs

```diff
@@ -164,6 +172,59 @@ public void Free()
             }
 
             _wrappers = null;
+            _constructorInitializers = null;
+        }
+
+        // Add an edge to the ctor-initializer graph
+        internal void NoteConstructorInitializer(MethodSymbol method1, MethodSymbol method2, CSharpSyntaxNode syntax)
+        {
+            if (_constructorInitializers == null)
+            {
+                _constructorInitializers = new Dictionary<MethodSymbol, Tuple<MethodSymbol, CSharpSyntaxNode>>();
+            }
+
+            _constructorInitializers.Add(method1, new Tuple<MethodSymbol, CSharpSyntaxNode>(method2, syntax));
+        }
+
+        // Report any cycles in the ctor-intializer graph starting at the given ctor.
+        internal void ReportCtorInitializerCycles(MethodSymbol ctor, DiagnosticBag diagnostics)
+        {
+            if (_constructorInitializers == null)
+            {
+                return;
+            }
+
+            var visiting = PooledHashSet<MethodSymbol>.GetInstance();
+            var todo = PooledHashSet<MethodSymbol>.GetInstance();
+            todo.AddAll(_constructorInitializers.Keys);
+
+            if (HasCycleAt(ctor, visiting, todo))
+            {
+                var errorLocation = _constructorInitializers[ctor].Item2.Location;
+                diagnostics.Add(ErrorCode.ERR_RecursiveConstructorCall2, errorLocation, ctor);
+                _constructorInitializers.Remove(ctor);
+            }
+
+            visiting.Free();
+            todo.Free();
+        }
+
+        private bool HasCycleAt(MethodSymbol ctor, HashSet<MethodSymbol> visiting, HashSet<MethodSymbol> todo)
```

Comment:
I believe it's possible for the current implementation to return true for a constructor which is not actually part of a recursive loop. For example, consider the following constructors used in the test:

```csharp
public A() : this(1) {}
public A(int x) : this(string.Empty) {}
public A(string s) : this(1) {}
```

Do we want to make sure this warning is only reported on `A.A(int)` or `A.A(string)`, or is it also acceptable for it to be reported on `A.A()`. Note that if we report it on `A.A()`, the use of `_constructorInitializers.Remove(ctor)` above to "break the loop" will not behave as expected, and you could see two errors for a single loop.